File: sdks/markets/ts/test/helpers/account-metas.ts
/**
 * Shared fixtures for the account-meta golden guard tests.
 *
 * Extracted verbatim from the original `account-metas.test.ts` so the
 * kassandra-market and MetaDAO golden suites can share the placeholder
 * addresses, the program-id table, and the `label()` helper without drift.
 */
import { Address, TransactionInstruction } from "@solana/web3.js";

import {
  EXTERNAL_PROGRAM_IDS,
  SYSTEM_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
} from "../../src/constants.js";
import { ATA_PROGRAM_ID } from "../../src/metadao/constants.js";

export const CV_PROGRAM = EXTERNAL_PROGRAM_IDS.conditionalVault;
export const AMM_PROGRAM = EXTERNAL_PROGRAM_IDS.ammV04;

/** Deterministic distinct placeholder address (32 bytes of `n`). */
export const A = (n: number): Address => new Address(new Uint8Array(32).fill(n));

/** One golden row: the account's role name + its signer/writable flags. */
export type Meta = [name: string, isSigner: boolean, isWritable: boolean];

/**
 * Label each account meta by looking its pubkey up in `entries` (a reverse map of
 * `address -> role`). Throws on any unmapped account so a builder that grew/moved a
 * slot fails loudly rather than silently. Returns `[role, isSigner, isWritable]`.
 */
export function label(ix: TransactionInstruction, entries: Array<[Address, string]>): Meta[] {
  const rev = new Map(entries.map(([a, n]) => [a.toString(), n]));
  return ix.keys.map((k) => {
    const name = rev.get(k.pubkey.toString());
    if (name === undefined) {
      throw new Error(`unmapped account ${k.pubkey.toString()} in ${ix.keys.length}-key ix`);
    }
    return [name, k.isSigner, k.isWritable];
  });
}

/** The fixed program-id accounts, by role. Spread into per-instruction maps. */
export const PROGRAMS: Array<[Address, string]> = [
  [SYSTEM_PROGRAM_ID, "systemProgram"],
  [TOKEN_PROGRAM_ID, "tokenProgram"],
  [ATA_PROGRAM_ID, "ataProgram"],
  [CV_PROGRAM, "cvProgram"],
  [AMM_PROGRAM, "ammProgram"],
];

// Distinct placeholder args reused across the kassandra-market builders.
export const PAYER = A(1);
export const KASS_MINT = A(2);
export const AUTHORITY = A(3);
export const ORACLE = A(4);
export const CREATOR = A(5);
export const CREATOR_ATA = A(6);
export const CONTRIBUTOR = A(7);
export const CONTRIB_ATA = A(8);
export const LP_ATA = A(9);
export const QUESTION = A(10);
export const FEE_DEST = A(11);
export const VAULT = A(12);
export const VAULT_UNDERLYING_ATA = A(13);
export const YES_MINT = A(14);
export const NO_MINT = A(15);
export const MARKET_CYES = A(16);
export const MARKET_CNO = A(17);
export const AMM = A(18);
export const LP_MINT = A(19);
export const LP_VAULT_ACC = A(20);
export const AMM_VAULT_BASE = A(21);
export const AMM_VAULT_QUOTE = A(22);
export const CV_EVENT_AUTH = A(23);
export const AMM_EVENT_AUTH = A(24);

File: sdks/markets/rust/tests/account_metas.rs
//! Account-meta golden guard — pins each instruction's ACCOUNT CONTRACT.
//!
//! The `tests/parity.rs` guard locks discriminators, sizes, and payload bytes,
//! but a wrong account ORDER or a flipped `is_signer`/`is_writable` flag would
//! still slip past CI. This file closes that gap: for every `ix::*` builder AND
//! every `metadao::*` account builder, it builds the instruction with DISTINCT
//! placeholder `Pubkey`s (deriving every PDA the builder derives), then asserts
//! the resulting `Vec<(role, is_signer, is_writable)>` equals a HARDCODED literal
//! golden — a hand-written frozen snapshot, NOT computed from the builder.
//!
//! Cross-checked against the program processors' `let [a, b, ..]` destructures
//! (`programs/markets/src/processor/*.rs`) and the CPI metas
//! (`activate.rs` split_metas, `collect_fee.rs` redeem_metas). Where the two SDKs
//! disagreed the PROGRAM wins; the labels + order here are IDENTICAL to the TS
//! golden in `sdks/oracles/ts/test/account-metas.test.ts`, so both SDKs encode ONE contract.
//! Any future account-order/flag drift in either SDK fails these tests.
//!
//! The `#[test]`s are split across sibling modules (`kassandra`, `metadao`); this
//! file keeps the shared placeholder/labeling helpers they build on.

#[path = "account_metas/kassandra.rs"]
mod kassandra;
#[path = "account_metas/metadao.rs"]
mod metadao;

use kassandra_markets_sdk::metadao as md;
use solana_sdk::{instruction::Instruction, pubkey::Pubkey};
use std::collections::HashMap;

/// A single golden row: role name + signer/writable flags.
pub(crate) type Meta = (&'static str, bool, bool);

/// Deterministic distinct placeholder pubkey (32 bytes of `n`).
pub(crate) fn pk(n: u8) -> Pubkey {
    Pubkey::new_from_array([n; 32])
}

/// Label each account meta by looking its pubkey up in `entries`. Panics on any
/// unmapped account so a builder that grew/moved a slot fails loudly.
pub(crate) fn labeled(ix: &Instruction, entries: Vec<(Pubkey, &'static str)>) -> Vec<Meta> {
    let map: HashMap<Pubkey, &'static str> = entries.into_iter().collect();
    ix.accounts
        .iter()
        .map(|m| {
            let name = map
                .get(&m.pubkey)
                .unwrap_or_else(|| panic!("unmapped account {} in {:?}", m.pubkey, ix.program_id));
            (*name, m.is_signer, m.is_writable)
        })
        .collect()
}

/// The fixed program-id accounts, by role. Spread into per-instruction maps.
fn programs() -> Vec<(Pubkey, &'static str)> {
    vec![
        (solana_sdk::system_program::id(), "systemProgram"),
        (spl_token::id(), "tokenProgram"),
        (md::ASSOCIATED_TOKEN_PROGRAM_ID, "ataProgram"),
        (md::CONDITIONAL_VAULT_ID, "cvProgram"),
        (md::AMM_ID, "ammProgram"),
    ]
}

pub(crate) fn with_programs(
    mut entries: Vec<(Pubkey, &'static str)>,
) -> Vec<(Pubkey, &'static str)> {
    entries.extend(programs());
    entries
}
